client-ci-analysis/ci-failure-analysis/Program.cs

```diff
@@ -0,0 +1,98 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace ci_failure_analysis
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            var jsonFiles = new[]
+            {
+                @"private_builds.json",
+                @"official_builds.json"
+            };
+
+            foreach (var (jsonPath, index) in jsonFiles.Select((item, index) => (item, index)))
+            {
+                Console.WriteLine(Path.GetFileName(jsonPath));
+
+                var jsonText = File.ReadAllText(jsonPath);
+                var builds = System.Text.Json.JsonSerializer.Deserialize<List<BuildInfo>>(jsonText);
+
+                var totalBuilds = builds.Count;
+                var totalFailures = builds.SelectMany(b => b.issues).Count();
+                var infrastructureFailures = builds.Where(b => b.issues.Any(r => !IsUserFailure(r))).SelectMany(b => b.issues).Count();
+
+                Console.WriteLine("Failed builds: " + totalBuilds);
+                Console.WriteLine("Failure reasons: " + totalFailures);
+                Console.WriteLine("Infrastructure failures: " + infrastructureFailures);
+                Console.WriteLine();
+
+                var buildsByFailures = builds.SelectMany(b => b.issues.Select(r => new { Build = b.issues, Reason = r })).GroupBy(b => b.Reason).ToDictionary(b => b.Key, b => b.Select(r => r.Build).ToList());
+
+                foreach (var failure in buildsByFailures.OrderByDescending(f => f.Value.Count))
+                {
+                    int count = failure.Value.Count;
+                    int failureRate = 100 * count / totalFailures;
+                    int? infraFailureRate = IsUserFailure(failure.Key)
+                        ? (int?)null
+                        : 100 * count / infrastructureFailures;
+
+                    Console.WriteLine($"{failure.Key}: {failure.Value.Count} {failureRate}% {infraFailureRate?.ToString() ?? "-"}%");
+                }
+
+                var macHangs = builds.Where(b => b.issues.Contains("mac tests hung")).ToList();
```

Comment:
After `find-builds-in-sprint` outputs the build list, I open every build, one by one, and manually analyse why the build failed and type in the reason. Being able to auto-detect the failure reason would be the single biggest time saver for me, but it will also require considerable effort to add rules to understand all the different types of failures that happen.

Reply:
I see..Yes, that's true. 
Searched and found we don't have a tracking issue for auto-detect the failure reason? 
I'll create one.

Reply:
https://github.com/NuGet/Client.Engineering/issues/361 created.